error-issue/src/error_monitor.ts

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright 2020 The AMP HTML Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {StackdriverApi} from './stackdriver_api';
+import {ErrorReport, Stackdriver} from './types';
+
+import fetch from 'node-fetch';
+import statusCodes from 'http-status-codes';
+
+const ERROR_ISSUE_REPORT_ENDPOINT =
+  'https://us-central1-amp-error-issue-bot.cloudfunctions.net/error-issue';
+
+export class ErrorMonitor {
+  constructor(
+    private client: StackdriverApi,
+    private minFrequency: number = 5000,
+    private pageLimit: number = 25
+  ) {}
+
+  /** Tests if an error group already has an associated issue. */
+  private hasTrackingIssue(group: Stackdriver.ErrorGroup) {
+    return !!group.trackingIssues;
+  }
+
+  /** Tests if an error group is occurring frequently enough to report. */
+  private hasHighFrequency(groupStats: Stackdriver.ErrorGroupStats) {
+    const timedCount = groupStats.timedCounts[0];
+    return timedCount && timedCount.count >= this.minFrequency;
+  }
+
+  /** Finds frequent errors to create tracking issues for. */
+  async newErrorsToReport(): Promise<Array<ErrorReport>> {
+    return (await this.client.listGroups(this.pageLimit))
+      .filter(({group}) => !this.hasTrackingIssue(group))
+      .filter(groupStats => this.hasHighFrequency(groupStats))
+      .map(groupStats => this.reportFromGroupStats(groupStats));
+  }
+
+  private reportFromGroupStats({
```

Comment:
JSDoc + I find it unsettling that the order of methods here is:
```typescript
  private a() {}
  private b() {}
  func() {
    a();
    b();
    c();
  }
  private c() {}
```

Reply:
Done.